### PR TITLE
Start building up better data structures for static cross-module info

### LINKED ISSUE #288: Repl can't tell the difference between identifiers and variables
At the REPL:

> var foo = 5
> foo := 5

Will report that foo was not defined as a variable.  This is because the repl exports don't distinguish correctly, so this is being statically flagged as an error before running the code at the repl.  The fix is to add more information in CompileEnv bindings so the compiler allows variables in its initial scope, and then have repl-support and repl-lib correctly add variables vs. identifiers to the environment.

### PR DESCRIPTION
**Pull request by jpolitz**
(@blerner – requesting review; I would actually like to push this forward if possible, since it fixes a longstanding issue with variables at the REPL, while also working with data structures required for flatness analysis.)

Most directly, this commit fixes #288.

It also gives a place to start hanging more static information off of module
exports.  The information is all "there" but needs to be attached to particular
bindings in different phases.

This should eventually help with statically checking cases (#366) and attaching
cross-module locations to bindings, or associating bindings with import lines
(#678).  The `ValueExport` type also captures the flatness analysis information
that @puppyofkosh is working on.

I'm not sure that the `ValueExport` datatype is in its final form, and how many
variants it will eventually have.  But it's a starting point that makes
variables work, and demonstrated how to get more than just type information
propagated around across modules.  In particular, it maybe should be unified
with the the `ScopeBinding` type, since they reflect similar information, just
cross- vs intra-module.

This commit also uses the "defined-values" field instead of the object-shaped
"values" in "provide-plus-types".  This is because, with variables exported,
it's impossible to represent an exported variable in a Pyret object, since a
Pyret object's field could only hold the value of the variable.  Here, the cell
itself is exported.

I think this is actually a surprisingly expressive feature for a module system
to offer, though the alternative is to rework how the REPL works entirely to
not use module imports and exports as the semantics of binding between entries.

### CODE REVIEW

File: src/arr/compiler/compile-structs.arr [hunk 1]
```diff
@@ -190,10 +210,9 @@ fun provides-from-raw-provides(uri, raw):
   values = raw.values
   vdict = for fold(vdict from SD.make-string-dict(), v from raw.values):
     if is-string(v) block:
-      vdict.set(v, t-top)
+      vdict.set(v, v-just-type(t-top))
     else:
-      #print("\n\nYOOOOOOOOOOOOOOOOOOOOOOO: " + tostring(v.name))
```
Comment (jpolitz) [inline src/arr/compiler/compile-structs.arr]: hi

**Comment by shriram** [comment]
Improving hanging static information off exports will also be relevant to
Ian Boros as he conveys information like call depth. We talked today about
maybe doing a more general abstract interpretation framework for lots of
other little optimizations.

**Comment by jpolitz** [comment]
cca25044213b8e5efa2ea1658a24a39bdaab1f64 had some changes to the exports of `sets`, because this commit is the last step in truly disconnecting `provide`'s semantics from "create an arbitrary Pyret value and shove it across the module boundary." Instead, `provide` can (currently) _only_ make available names that are defined at the top level of a module.  Their values will be whatever the value of those names are in scope at the very end of the module, hence the new uses of `shadow` at the end of `sets.arr`.

The way forward here is to enrich `provide` syntax with primitives like Racket's `rename-out`, `prefix-out`, etc, to control these names reasonably.  But getting away from "just make a value" is the first step, since now Pyret is developing a really good model for cross-module bindings in a first-order way, something that's much harder to reconstruct from an object value.

**Comment by jpolitz** [comment]
Also, I think this is ready to merge with this commit once tests pass, modulo any cleanup.  There are no "dangling" parts of the implementation.  Variable exports work for both serialized and in-memory modules, and at the REPL, which is a complete implementation of static/dynamic information that isn't just vanilla values and types, so we can build on this going forward.

It also makes variables work at the REPL, closing an issue from 2014 that doesn't likely have a simpler resolution :-)

**Pull request merged**